### ISSUE TITLE
Early stopping with mmdetection

### ISSUE THREAD

**Issue by EtienneDavid** [open-mmlab/mmdetection#2062]
Dear all,

Thank again for the amazing library. It's a real pleasure to work with it.

I would like to introduce early stopping conditions in my training schedule, is it something which is already implement ? If not, could you point me to the modules that would need to be updated for such features ?

Best

**Comment by ZwwWayne** [comment]
Hi @EtienneDavid ,
We have not implemented it because it is a little bit rarely used in the detection field.
Since it is about the training pipeline, it is worth trying to implement it as a hook or something used by the [runner in mmcv](https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/runner.py).

**Comment by EtienneDavid** [comment]
OK thank you for your answer. Why is it rarely used in detection field ? Is there an habit or there are more theorical reasons to have a fixed schedule ?

**Comment by ZwwWayne** [comment]
Practically, the overfitting phenomenon with the COCO dataset is not so apparent in common schedules (1x, 20e, 2x) since the dataset is large and augmentations are already enough.

**Comment by EtienneDavid** [comment]
OK thank you very much :)

**Comment by xincyu** [comment]
But it becomes useful when training to customize a small number of data sets

**Comment by Prateek-Havanur** [comment]
@EtienneDavid @xincyu Hello Guys i am trying to implement early stopping mmdetection. it would be of great help if you can provide any insights apart from already mentioned. Since i am using very small dataset its needed. As of now i am running workflow with [(train,1)(val,1)] to capture losses on val set. Any help would be very much appreciated :)

**Comment by FrancescoSaverioZuppichini** [comment]
Same problem, early stopping is a **must** when finetuning a model. I've tried to think how to implement it in mmdetection but it is so hard to extend the system and fetch the metric that it may be impossible without some heavy hacks

**Comment by fzoric8** [comment]
Early stopping would be really useful when someone wants to finetune a model to the small dataset.

**Comment by leetoffolo** [comment]
[EarlyStoppingHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/early_stopping_hook.py) is available since mmengine 0.7.0. (MMDetection >= 3.0) 
You can call the hook it in your config under the "default_hooks" dictionary:
`default_hooks = dict(
    early_stopping=dict(type='EarlyStoppingHook',
        monitor="mAP",
        patience=10,
        min_delta=0.005),
    checkpoint=dict(
        type="CheckpointHook",
        interval=interval,
        save_begin=100,
        max_keep_ckpts=max_keep_ckpts,
        save_best=save_best)
)`

**Comment by animatedbrick** [comment]
Just expanding on @leetoffolo answer, is there a way to have dynamic learning rate schedules? Ie if coco/bbox_mAP does not increase for 10 epochs by min delta, drop the learning rate by a factor of 10 rather than finish training. I would like to do this twice, ie 2 learning rate drops, and the next time min delta change is not achieved finish the training. 

I am comparing pretraining weights, ie COCO vs Imagenet. However, I have noticed that models pre-trained on COCO converge faster than Imagenet (this is expected from literature and transfer learning). However, models pre-trained on Imagenet are performing better than their COCO counterparts, I am attributing this to the Imagenet models running long enough to utilize the learning rate scheduler, however, the COCO models converge and finish training before the learning rate is dropped. 

I guess a simple solution may be to decrease min delta & increase patience to ensure the models train long enough to hit the learning rate schedules, but this wouldn't be the greatest approach.
